ledger/README.md

```diff
@@ -7,3 +7,58 @@ This package provides two ledger implementations:
 - **Complete Ledger** implements a fast, memory-efficient and reliable ledger. It holds a limited number of recently used states in memory (for speed) and uses write-ahead logs and checkpointing to provide reliability. Under the hood complete ledger uses a collection of MTries(forest). MTrie is a customized in-memory binary Patricia Merkle trie storing payloads at specific storage paths. The payload includes both key-value pair and storage paths are determined by the PathFinder. Forest utilizes unchanged sub-trie sharing between tries to save memory.
 
 - **Partial Ledger** implements the ledger functionality for a limited subset of keys. Partial ledgers are designed to be constructed and verified by a collection of proofs from a complete ledger. The partial ledger uses a partial binary Merkle trie which holds intermediate hash value for the pruned branched and prevents updates to keys that were not part of proofs.
+
+## Definitions
+
+### binary Merkle tree
+
+![binary Merkle tree image](/ledger/docs/binary_merkle_tree.png?raw=true "binary Merkle tree" )
+
+In this context a *binary Merkle tree* is defined as a full binary tree with an specific height including three type of nodes:
+
+- leaf nodes: holds a payload (data), a path (where the node is located), and a hash value (hash of path and payload content)
+
+- empty leaf nodes: doesn't hold any data and only stores a path, and a default hash value based on the height of tree
+
+- intermediate nodes: holds a path and a hash value which is defined as hash of hash value of left and right children.
+
+![node types image](/ledger/docs/node_types.png)
+
+A *path* is a unique address of a node storing a payload. Paths are derived from the content of payloads (see common/pathfinder).
+
+![paths image](/ledger/docs/paths.png?raw=true "paths")
+
+#### Operations
+
+**Get**: Fetching a payload from the binary Merkle tree is by traversing the tree based on path bits. (0: left branch, 1: right branch)
+
+**Update**: Updates to the tree starts with traversing the tree to the leaf node, updating payload, hash value of that node and hash value of all the ancestor nodes (nodes on higher level connected to this node).
+
+![update image](/ledger/docs/tree_update.gif?raw=true "update")
+
+**Prove**: A binary Merkle tree can provide an inclusion proof for any given payload. A Merkle proof in this context includes all the information needed to walk through a tree branch from an specific leaf node (key) up to the root of the tree (yellow node hash values are needed for inclusion proof for the green node).
+
+![proof image](/ledger/docs/proof.png?raw=true "proof")
+
+### binary Merkle trie
+A **binary Merkle trie** in this context is defined as a compact version of binary Merkle tree, providing exact same functionality but don’t store unnecessary nodes.
```

Comment:
I think it would be good to specify what "unnecessary nodes" are:

```suggestion
A **binary Merkle trie** in this context is defined as a compact version of binary Merkle tree, providing exact same functionality but doesn't explicitly store empty nodes. Formally, a node is empty: 
* the node is an empty leaf node (see definition above)
* an interior node is defined to be empty, if its two children are empty. 
```

Reply:
updated at 0d1658b